### PR TITLE
SHS-NG M4.3: Port StorageTab to the new backend.

### PR DESCRIPTION
**Pull request by vanzin**
This required adding information about StreamBlockId to the store,
which is not available yet via the API. So an internal type was added
until there's a need to expose that information in the API.

The UI only lists RDDs that have cached partitions, and that information
wasn't being correctly captured in the listener, so that's also fixed,
along with some minor (internal) API adjustments so that the UI can
get the correct data.

Because of the way partitions are cached, some optimizations w.r.t. how
often the data is flushed to the store could not be applied to this code;
because of that, some different ways to make the code more performant
were added to the data structures tracking RDD blocks, with the goal of
avoiding expensive copies when lots of blocks are being updated.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -533,17 +538,22 @@ private[spark] class AppStatusListener(
               dist.onHeapRemaining = newValue(dist.onHeapRemaining, -memoryDelta)
             }
           }
+          dist.lastUpdate = null
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: I think [my prev comment on exec memory remaining with multiple rdds](https://github.com/apache/spark/pull/19383#discussion_r146689668) still applies here.  you're only updating it for the one RDD with an update, but the memory remaining is actually changing for all rdds cached on this executor

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 2]
```diff
@@ -533,17 +538,22 @@ private[spark] class AppStatusListener(
...
       rdd.memoryUsed = newValue(rdd.memoryUsed, memoryDelta)
       rdd.diskUsed = newValue(rdd.diskUsed, diskDelta)
-      update(rdd)
+
+      // Need to always flush the RDD in live applications in case there's a single cached block,
+      // which means there won't be a subsequent event that could trigger the flush.
+      if (live) {
+        update(rdd)
+      } else {
+        liveUpdate(rdd)
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: isn't this whole if/else just `liveUpdate(rdd)`?

File: core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala [hunk 3]
```diff
@@ -725,10 +730,33 @@ class AppStatusListenerSuite extends SparkFunSuite with BeforeAndAfter {
 
     // Unpersist RDD1.
     listener.onUnpersistRDD(SparkListenerUnpersistRDD(rdd1b1.rddId))
-        intercept[NoSuchElementException] {
+    intercept[NoSuchElementException] {
       check[RDDStorageInfoWrapper](rdd1b1.rddId) { _ => () }
     }
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/status/AppStatusListenerSuite.scala:767]: so I don't forget, repeating the comment for adding the test for multiple rdds on one executor

**Review by squito — COMMENTED** [review]
I haven't looked at StreamBlocks before, and I have to say I'm totally perplexed even by the current implementation.  It looks to me like its just every block on the executors, and nothing to do with streaming at all (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala#L94).  I will look at that more.  But what you have here seems consistent with that anyway

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala (hunk 2 above)
Comment (vanzin) [reply core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: The `update` call needs an extra argument here...

**Pull request closed without merging**